File: java-basic/src/main/java/ch15/Test05.java
// Object 클래스 - equals() 오버라이딩 
package ch15;

class My5 {
  String name;
  int age;
  
  @Override
  public boolean equals(Object obj) {
    if (this == obj)
      return true;
    if (obj == null)
      return false;
    if (getClass() != obj.getClass())
      return false;
    My5 other = (My5) obj;
    if (age != other.age)
      return false;
    if (name == null) {
      if (other.name != null)
        return false;
    } else if (!name.equals(other.name))
      return false;
    return true;
  }
  
  
}

public class Test05 {
  public static void main(String[] args) {
    My5 obj1 = new My5();
    obj1.name = "홍길동";
    obj1.age = 20;
    
    My5 obj2 = new My5();
    obj2.name = "홍길동";
    obj2.age = 20;
    
    System.out.println(obj1 == obj2);
    System.out.println(obj1.equals(obj2));

    // 결론!
    // => Object에서 상속 받은 것을 그대로 사용하면 equals()는 인스턴스가 같은지 비교한다.
    // => 인스턴스의 내용물이 같은지 비교하도록 만들고 싶다면 equals()을 오버라이딩 하라!
    // => String와 wrapper 클래스는 equals() 오버라이딩 하였다.
    // => StringBuffer 클래스는 equals()를 오버라이딩 하지 않았다.
  }

}

File: node-basic/step03/test05.js
/* 주제: 흐름 제어문 - 조건문 if ~ else if ~ else if ~ else
- 문법
if (식)   <-- 식의 결과가 참이면 '명령문1'이 실행되고, 거짓이면 else 가 실행된다.
  명령문1;
else
  if (식) <-- 식의 결과가 참이면 '명령문2'가 실행되고, 거짓이면 else 가 실행된다.
    명령문2
  else
    명령문3
*/
"use strict"

var age;

age = 11;
if (age < 8)
  console.log("유아");
else {// <-- age >= 8
  if (age >= 8 && age < 14)
    console.log("어린이");
  else  // <-- age >= 14
    console.log("청소년 이상입니다.");
}
console.log("----------------------");

// if ~ else ~는 한 그룹으로 취급한다.
age = 40;
if (age < 8)
  console.log("유아");
else // <-- age >= 8, if ~ else ~ 는 한 문장이기 때문에 블록을 생략해도 된다.
  if (age >= 8 && age < 14)
    console.log("어린이");
  else  // <-- age >= 14
    console.log("청소년 이상입니다.");
console.log("----------------------");

age = 65;
if (age < 8)
  console.log("유아");
else // <-- age >= 8, if ~ else ~ 는 한 문장이기 때문에 블록을 생략해도 된다.
  if (age >= 8 && age < 14)
    console.log("어린이");
  else  // <-- age >= 14
    if (age >= 14 && age < 20)
      console.log("청소년");
    else
      if (age >= 20 && age < 65)
        console.log("성인");
      else
        console.log("노인");
console.log("----------------------");

// 실무에서는 if ~ else ~ 문장을 보기 좋게 정렬한다.
// 주의!
// - if ~ else if ~ else if ~ else 라는 문법은 없다!
// - 그냥 if ~ else 코드를 정돈한 것이다.
age = 65;
if (age < 8)
  console.log("유아");
else if (age >= 8 && age < 14)
  console.log("어린이");
else if (age >= 14 && age < 20)
  console.log("청소년");
else if (age >= 20 && age < 65)
  console.log("성인");
else
  console.log("노인");
console.log("----------------------");


//
